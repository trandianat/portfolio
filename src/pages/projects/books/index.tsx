import { ChangeEvent, useEffect, useState } from 'react';
import { Link } from 'components/link';
import * as styles from 'pages/projects/books/styles';
import { nyTimesConfig } from 'utils/constants';

export const Books = (): JSX.Element => {
  const [allBooks, setAllBooks] = useState<{ [x: string]: string }[]>([]);
  const [books, setBooks] = useState<{ [x: string]: string }[]>([]);
  const [category, setCategory] = useState<string>('');
  const [categories, setCategories] = useState<Record<string, string>>({});
  const [date, setDate] = useState<string>('');
  const [dateChanged, setDateChanged] = useState<boolean>(false);
  const [defaultDate, setDefaultDate] = useState<string>('');
  const [message, setMessage] = useState<string>('');

  const duration = (weeks: number): string => {
    if (weeks > 52) {
      const years = Math.floor(weeks / 52);
      const weeksLeft = weeks - years * 52;
      return `${years} year${years === 1 ? '' : 's'}${
        weeksLeft > 0
          ? ` and ${weeksLeft} week${weeksLeft === 1 ? '' : 's'}`
          : ''
      }`;
    }
    return `${weeks} week${weeks === 1 ? '' : 's'}`;
  };

  useEffect(() => {
    if (category) {
      setBooks(
        allBooks.filter(
          ({ category_encoded }: Record<string, string>) =>
            category === category_encoded
        )
      );
    }
  }, [category]);

  useEffect(() => {
    if (date) {
      fetch(
        `${nyTimesConfig.apiUrl}?api-key=${nyTimesConfig.apiKey}&published_date=${date}`
      )
        .then(response => response.json())
        .then(({ results }) => {
          if (results !== undefined && Object.keys(results).length) {
            setCategories({});
            const sortedCategories = results.lists
              .map((list: Record<string, unknown>) => [
                list.list_name_encoded,
                list.display_name,
              ])
              .sort((a: string, b: string) => {
                if (a[1] < b[1]) {
                  return -1;
                } else if (a[1] > b[1]) {
                  return 1;
                } else {
                  return 0;
                }
              });
            sortedCategories.forEach(([key, value]: string[]) =>
              setCategories(categories => ({ ...categories, [key]: value }))
            );
            const bookResults: { [x: string]: string }[] = [];
            results.lists.forEach(
              (list: {
                books: { [x: string]: string }[];
                list_name_encoded: string;
              }) =>
                bookResults.push(
                  ...list.books.map((book: Record<string, unknown>) => ({
                    ...book,
                    category_encoded: list.list_name_encoded,
                  }))
                )
            );
            setAllBooks(bookResults);
            setBooks(bookResults);
          } else {
            setAllBooks([]);
            setBooks([]);
            setCategories({});
            setMessage('No results found for the selected publication date');
          }
        });
    }
  }, [date]);

  useEffect(() => {
    const today = new Date();
    const year = today.getFullYear();
    const month = today.getMonth() + 1;
    const day = today.getDate();
    const formattedMonth = month < 10 ? '0' + month : month;
    const formattedDay = day < 10 ? '0' + day : day;
    const formattedDate = `${year}-${formattedMonth}-${formattedDay}`;
    setDefaultDate(formattedDate);
    setDate(formattedDate);
  }, []);

  return (
    <div css={styles.books}>
      <h2>Books</h2>
      <p>
        Search for New York Times' best-selling books by the date that the
        best-seller list was published and by category. Best-seller lists are
        published each week, so the selected date will correspond to the list
        published for that week.
      </p>
      <div className="search">
        <div className="filter">
          <label htmlFor="id">Publication date:</label>
          <input
            defaultValue={defaultDate}
            id="date"
            min="2008-06-01"
            max={defaultDate}
            onChange={({
              target: { value },
            }: ChangeEvent<HTMLInputElement>) => {
              setCategory('');
              setMessage('');
              (
                document.getElementById('category') as HTMLSelectElement
              ).selectedIndex = 0;
              const validDate =
                /(19|20)\d\d-(0[1-9]|1[0-2])-(0[1-9]|[12][0-9]|3[01])/;
              if (validDate.test(value)) {
                setDate(value);
              } else {
                setDate('');
                setCategories({});
                setMessage('Invalid publication date');
              }
              if (!dateChanged) {
                setDateChanged(true);
              }
            }}
            type="date"
          />
        </div>
        <div className="filter">
          <label htmlFor="category">Category:</label>
          <select
            defaultValue=""
            id="category"
            onChange={(event: ChangeEvent<HTMLSelectElement>) => {
              setBooks(allBooks);
              setCategory(event.target.value);
            }}
            {...(!Object.keys(categories).length && { disabled: true })}
          >
            <option disabled value="">
              Select a category
            </option>
            {Object.entries(categories).map(([key, value]: string[]) => (
              <option key={key} value={key}>
                {value}
              </option>
            ))}
          </select>
        </div>
        <button
          onClick={() => {
            setCategory('');
            setDate(defaultDate);
            setDateChanged(false);
            setMessage('');
            (document.getElementById('date') as HTMLInputElement).value =
              defaultDate;
            (
              document.getElementById('category') as HTMLSelectElement
            ).selectedIndex = 0;
          }}
          {...(!dateChanged && !category && { disabled: true })}
        >
          Reset
        </button>
      </div>
      {category && (
        <div className="books">
          {books.map(
            ({
              author,
              book_image,
              buy_links,
              category_encoded,
              description,
              primary_isbn13,
              rank,
              title,
              weeks_on_list,
            }: {
              [x: string]: any;
            }) => (
              <div
                className="book"
                key={`${category_encoded}-${primary_isbn13}`}
              >
                <img src={book_image} />
                <div className="details">
                  <div className="header">
                    <div className="title">
                      <h3>{title}</h3>
                      <p>{author}</p>
                      {weeks_on_list > 0 && (
                        <p>{duration(weeks_on_list)} on best-seller list</p>
                      )}
                    </div>
                    <div className="rank">{rank}</div>
                  </div>
                  {(description || buy_links) && <hr />}
                  {description && <p>{description}</p>}
                  <div className="purchase">
                    <p>Purchase:</p>
                    {buy_links.map((link: Record<string, string>) => (
                      <Link key={link.name} text={link.name} url={link.url} />
                    ))}
                  </div>
                </div>
              </div>
            )
          )}
        </div>
      )}
      {message && <p className="message">{message}</p>}
    </div>
  );
};
