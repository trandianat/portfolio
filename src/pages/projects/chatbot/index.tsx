import { LexRuntimeService } from '@aws-sdk/client-lex-runtime-service';
import { ChangeEvent, useEffect, useState } from 'react';
import { v4 } from 'uuid';
import * as styles from 'pages/projects/chatbot/styles';
import { Persona } from 'utils/constants';

type Message = {
  from: Persona.AUTO | Persona.BOT | Persona.USER;
  input: string;
};

export const Chatbot = (): JSX.Element => {
  const [error, setError] = useState<boolean>(false);
  const [input, setInput] = useState<string>('');
  const [messages, setMessages] = useState<Message[]>([]);

  const empty = input.length < 1;
  const whitespace = input.length > 0 && !input.trim();

  const lex = new LexRuntimeService({
    credentials: {
      accessKeyId: process.env.REACT_APP_LEX_ACCESS as string,
      secretAccessKey: process.env.REACT_APP_LEX_SECRET as string,
    },
    region: process.env.REACT_APP_REGION,
  });

  const handleError = () => {
    setError(true);
    setMessages([
      ...messages,
      {
        from: Persona.AUTO,
        input: 'Chatbot error detected. Please try again later.',
      },
    ]);
  };

  const send = async () => {
    try {
      await lex.postText(
        {
          botAlias: process.env.REACT_APP_LEX_ALIAS,
          botName: process.env.REACT_APP_LEX_BOT,
          inputText: input,
          userId: process.env.REACT_APP_LEX_USER,
        },
        (_: Error, data: any) => {
          console.log(data);
          if (data) {
            if (data.dialogState === 'ReadyForFulfillment') {
              setMessages([
                ...messages,
                { from: Persona.USER, input },
                {
                  from: Persona.BOT,
                  input: `Your ${data.slots.AppointmentType} appointment has been scheduled on ${data.slots.Date} at ${data.slots.Time}.`,
                },
              ]);
            } else {
              setMessages([
                ...messages,
                { from: Persona.USER, input },
                { from: Persona.BOT, input: data.message },
              ]);
            }
          } else {
            handleError();
          }
        }
      );
      setInput('');
    } catch (error) {
      handleError();
    }
  };

  useEffect(() => {
    const history = document.getElementById('messages') as HTMLElement;
    history.scrollTop = history.scrollHeight;
  }, [messages]);

  return (
    <div css={styles.chatbot}>
      <h2>Chatbot</h2>
      <p>
        This is an <b>Amazon Lex</b> chatbot that schedules appointments. The
        chatbot uses <b>machine learning</b> to process user input and fulfill
        three "slots"—appointment type, date, and time—before confirming an
        appointment. Feel free to chat with it and see how it responds.
      </p>
      <div css={styles.conversation}>
        <div css={styles.messages} id="messages">
          {!messages.length && <h3>Try asking for an appointment!</h3>}
          {messages.length > 0 &&
            messages.map(({ from, input }: Message) => (
              <div className={`message ${from}`} key={v4()}>
                {input}
              </div>
            ))}
        </div>
        <div css={styles.inputs}>
          <input
            disabled={error}
            maxLength={100}
            onChange={(event: ChangeEvent<HTMLInputElement>) =>
              setInput(event.target.value)
            }
            onKeyDown={event => {
              if (event.key === 'Enter') {
                document.getElementById('send')?.click();
              }
            }}
            placeholder="Type a message..."
            value={input}
          />
          <div className="buttons">
            <button
              disabled={empty || error || whitespace}
              id="send"
              onClick={() => send()}
            >
              Send
            </button>
            <button
              onClick={() => {
                setError(false);
                setInput('');
                setMessages([]);
              }}
            >
              Reset
            </button>
          </div>
        </div>
      </div>
    </div>
  );
};
