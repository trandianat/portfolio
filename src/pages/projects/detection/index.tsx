import { ChangeEvent, useEffect, useMemo, useState } from 'react';
import '@tensorflow/tfjs-backend-cpu';
import '@tensorflow/tfjs-backend-webgl';
import * as cocoSsd from '@tensorflow-models/coco-ssd';
import dog from 'assets/images/dog.jpg';
import street from 'assets/images/street.jpg';
import table from 'assets/images/table.jpg';
import wine from 'assets/images/wine.jpg';
import * as styles from 'pages/projects/detection/styles';

type Prediction = {
  bbox: number[];
  class: string;
  score: number;
};

export const Detection = (): JSX.Element => {
  const colors = useMemo(
    () => [
      'red',
      'blue',
      'green',
      'orange',
      'purple',
      'gray',
      'pink',
      'brown',
      'black',
      'yellow',
    ],
    []
  );
  const [error, setError] = useState(false);
  const [image, setImage] = useState(dog);
  const [loading, setLoading] = useState(false);
  const [results, setResults] = useState<Prediction[]>([]);

  const vowels = ['a', 'e', 'i', 'o', 'u'];

  const images: { [image: string]: string } = {
    dog: dog,
    street: street,
    table: table,
    wine: wine,
  };

  const detectImage = async () => {
    setLoading(true);
    const model = await cocoSsd.load();
    try {
      const predictions = await model.detect(
        document.getElementById('image') as HTMLImageElement,
        10,
        0.25
      );
      setResults(predictions as Prediction[]);
    } catch {
      setError(true);
    } finally {
      setLoading(false);
    }
  };

  useEffect(() => {
    const canvas = document.getElementById('canvas') as HTMLCanvasElement;
    const context = canvas.getContext('2d') as CanvasRenderingContext2D;
    if (results.length) {
      setError(false);
      results.forEach((result: Prediction, index: number) => {
        context.beginPath();
        context.lineWidth = 2;
        context.strokeStyle = colors[index];
        const [x, y, width, height] = result['bbox'];
        context.strokeRect(x, y, width, height);
      });
    } else {
      const selectedImage = document.getElementById(
        'image'
      ) as HTMLImageElement;
      context.clearRect(0, 0, selectedImage.width, selectedImage.height);
    }
  }, [colors, results]);

  return (
    <div css={styles.detection}>
      <h2>Detection</h2>
      <p>
        This project identifies objects within images via{' '}
        <b>TensorFlow's COCO-SSD machine learning object detection model</b>.
        The model returns a list of predicted objects, their coordinates within
        the image, and a score of confidence for each object. Results will vary
        per detection depending on what the model returns.
      </p>
      <div className="selection">
        <label htmlFor="dropdown">Select an image:</label>
        <select
          id="dropdown"
          onChange={(event: ChangeEvent<HTMLSelectElement>) => {
            setImage(event.target.value);
            setError(false);
            setResults([]);
          }}
          {...{ disabled: loading }}
        >
          {Object.keys(images).map((image: string) => (
            <option key={image} value={images[image]}>
              {image}
            </option>
          ))}
        </select>
      </div>
      {(results.length === 0 || error) && (
        <button onClick={() => detectImage()} {...{ disabled: loading }}>
          {loading ? 'Loading...' : 'Detect'}
        </button>
      )}
      <div className="result">
        <img id="image" src={image} />
        <canvas height={200} id="canvas" />
      </div>
      {error && (
        <p className="error text">
          Error in object detection. Please try again.
        </p>
      )}
      {results.length > 0 && (
        <div className="confidence">
          {results.map((result: Prediction, index: number) => (
            <p className="text" key={index}>
              A{vowels.includes(result.class.charAt(0)) ? 'n' : ''}{' '}
              <span style={{ color: colors[index] }}>
                <strong>{result.class}</strong>
              </span>{' '}
              with {Math.round(result.score * 100)}% confidence
            </p>
          ))}
        </div>
      )}
    </div>
  );
};
