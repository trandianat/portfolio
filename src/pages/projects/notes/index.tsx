import { API } from 'aws-amplify';
import { useContext, useEffect, useRef, useState } from 'react';
import Trash from 'assets/icons/trash';
import { createNote, deleteNote, updateNote } from 'graphql/mutations';
import { listNotes } from 'graphql/queries';
import * as styles from 'pages/projects/notes/styles';
import { Status } from 'utils/constants';
import { NotificationContext } from 'utils/notification';

type Note = {
  createdAt: string;
  description: string;
  id: string;
  name: string;
  updatedAt: string;
  __typename: string;
};

export const Notes = (): JSX.Element => {
  const { notify } = useContext(NotificationContext);
  const [notes, setNotes] = useState<Note[]>([]);
  const [title, setTitle] = useState<{ [title: string]: string }>({});
  const [content, setContent] = useState<{ [content: string]: string }>({});

  const locale = 'en-US';

  const useOutsideClick = (callback: Function) => {
    const ref = useRef<any>(null);
    useEffect(() => {
      const handleClick = (event: any) => {
        const titleClicked = event.target.innerText === ref.current?.value; // When h3 becomes input
        const contentClicked =
          event.target.innerText === ref.current?.innerText; // When p becomes textarea
        const inputClicked = event.target === ref.current; // When input or textarea is clicked
        if (!titleClicked && !contentClicked && !inputClicked) {
          callback();
        }
      };
      document.addEventListener('click', handleClick);
      return () => {
        document.removeEventListener('click', handleClick);
      };
    }, [ref]);
    return ref;
  };

  const handleClickOutside = () => {
    setTitle({});
    setContent({});
  };

  const ref = useOutsideClick(handleClickOutside);

  const getNotes = async () => {
    const {
      data: {
        listNotes: { items },
      },
    } = (await API.graphql({ query: listNotes })) as any;
    const dates = items.map((item: Note) => item.updatedAt);
    const sortedDates = dates.sort(
      (a: string, b: string) => Date.parse(b) - Date.parse(a)
    );
    const sortByUpdated = sortedDates.map((date: string) =>
      items.find((item: Note) => item.updatedAt === date)
    );
    setNotes(sortByUpdated);
  };

  const addNote = async (name: string, description: string) => {
    try {
      await API.graphql({
        query: createNote,
        variables: { input: { description, name } },
      });
      notify({ message: <p>New note added</p>, variant: Status.SUCCESS });
      getNotes();
    } catch {
      notify({ message: <p>Failed to add new note</p>, variant: Status.ERROR });
    }
  };

  const editNote = async ({ description, id, name }: Partial<Note>) => {
    try {
      await API.graphql({
        query: updateNote,
        variables: {
          input: {
            id,
            ...(name && { name }),
            ...(description && { description }),
          },
        },
      });
      notify({ message: <p>Note saved</p>, variant: Status.SUCCESS });
      getNotes();
    } catch {
      notify({ message: <p>Failed to save note</p>, variant: Status.ERROR });
    }
  };

  const removeNote = async (id: string) => {
    try {
      await API.graphql({
        query: deleteNote,
        variables: { input: { id } },
      });
      notify({ message: <p>Note removed</p>, variant: Status.SUCCESS });
      getNotes();
    } catch {
      notify({ message: <p>Failed to remove note</p>, variant: Status.ERROR });
    }
  };

  useEffect(() => {
    getNotes();
  }, []);

  return (
    <div css={styles.notes}>
      <h2>Notes</h2>
      <p>
        The notes are stored in <b>DynamoDB</b> and are added, edited, removed,
        and retrieved via <b>GraphQL</b>. You can click on the title or content
        of a note to edit it. The notes are sorted by latest update date, so
        notes are moved to the top of the list when they are updated. Feel free
        to add a note with suggestions for improvement on any of my projects!
      </p>
      <div>
        <p className="serif-italic">In progress:</p>
        <ul>
          <li>
            When clicking into a note's content, place the cursor at the end of
            the input (like how the cursor is placed at the end of the title)
          </li>
          <li>
            When deleting all of the text in a title or content, do not exit the
            input field
          </li>
          <li>Replace TypeScript "any" types</li>
          <li>Slide updated notes to the top of the list</li>
          <li>Error handling</li>
        </ul>
      </div>
      <button onClick={() => addNote('Untitled', 'Edit note')}>Add note</button>
      <div css={styles.cards}>
        {notes.length ? (
          notes.map(({ createdAt, description, id, name, updatedAt }: Note) => {
            const createdDate = new Date(createdAt);
            const updatedDate = new Date(updatedAt);
            const formattedCreatedDate = `${createdDate.toLocaleDateString(
              locale
            )} ${createdDate.toLocaleTimeString(locale)}`;
            const formattedUpdatedDate = `${updatedDate.toLocaleDateString(
              locale
            )} ${updatedDate.toLocaleTimeString(locale)}`;
            return (
              <div css={styles.card} key={id}>
                <div
                  className="trash"
                  onClick={() => {
                    removeNote(id);
                    setTitle({});
                    setContent({});
                  }}
                >
                  <Trash />
                </div>
                {title[id] ? (
                  <input
                    autoFocus
                    className="title"
                    id={id}
                    onBlur={() => {
                      if (title[id] !== name) {
                        editNote({ id, name: title[id] });
                      }
                      setTitle({});
                    }}
                    onChange={({ target: { value } }) =>
                      setTitle({ ...title, [id]: value })
                    }
                    onKeyDown={(event: any) => {
                      if (event.key === 'Enter') {
                        document.getElementById(id)?.blur();
                      }
                    }}
                    ref={ref}
                    value={title[id]}
                  />
                ) : (
                  <h3
                    className="name"
                    onClick={() => setTitle({ ...title, ...{ [id]: name } })}
                    ref={ref}
                  >
                    {name}
                  </h3>
                )}
                {content[id] ? (
                  <textarea
                    autoFocus
                    id={id}
                    onBlur={() => {
                      if (content[id] !== description) {
                        editNote({ id, description: content[id] });
                      }
                      setContent({});
                    }}
                    onChange={({ target: { value } }) =>
                      setContent({ ...content, [id]: value })
                    }
                    onKeyDown={(event: any) => {
                      if (event.key === 'Enter') {
                        document.getElementById(id)?.blur();
                      }
                    }}
                    ref={ref}
                    rows={4}
                    value={content[id]}
                  />
                ) : (
                  <p
                    className="description"
                    onClick={() =>
                      setContent({ ...content, ...{ [id]: description } })
                    }
                    ref={ref}
                  >
                    {description}
                  </p>
                )}
                <div css={styles.footer}>
                  <p>
                    <span>Created</span>: {formattedCreatedDate}
                  </p>
                  <p>
                    <span>Updated</span>: {formattedUpdatedDate}
                  </p>
                </div>
              </div>
            );
          })
        ) : (
          <div className="serif-italic" css={styles.empty}>
            No notes found
          </div>
        )}
      </div>
    </div>
  );
};
