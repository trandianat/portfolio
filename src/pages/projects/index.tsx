import { Link, Outlet, useLocation } from 'react-router-dom';
import Book from 'assets/icons/book';
import Chat from 'assets/icons/chat';
import Letter from 'assets/icons/letter';
import Note from 'assets/icons/note';
import Photo from 'assets/icons/photo';
import Wedding from 'assets/icons/wedding';
import Auth from 'components/auth';
import { Content } from 'components/content';
import * as styles from 'pages/projects/styles';
import isProjects from 'utils/projects';

type Project = {
  description: string;
  icon: JSX.Element;
  name: string;
  path: string;
};

export const Projects = (): JSX.Element => {
  const { pathname } = useLocation();
  const projects: Project[] = [
    {
      description:
        'Create, read, update, and delete notes sorted by latest update',
      icon: <Note />,
      name: 'Notes',
      path: 'notes',
    },
    {
      description: 'Detect objects in images by percentages of confidence',
      icon: <Photo />,
      name: 'Detection',
      path: 'detection',
    },
    {
      description:
        'Conversational artificial intelligence for scheduling appointments',
      icon: <Chat />,
      name: 'Chatbot',
      path: 'chatbot',
    },
    {
      description:
        'A website for my wedding guests to RSVP and find more information',
      icon: <Wedding />,
      name: 'Wedding',
      path: 'wedding',
    },
    {
      description:
        "Search for New York Times' best-selling books by date and category",
      icon: <Book />,
      name: 'Books',
      path: 'books',
    },
    {
      description: 'A recreation of the popular New York Times word game',
      icon: <Letter />,
      name: 'Wordle',
      path: 'wordle',
    },
  ];
  return (
    <Content>
      <div css={styles.auth}>
        <Auth>
          {isProjects(pathname) && (
            <div css={styles.projects}>
              {projects.map(({ description, icon, name, path }: Project) => (
                <Link css={styles.card} key={name} to={path}>
                  {icon}
                  <div css={styles.project}>
                    <h2>{name}</h2>
                    <p className="description">{description}</p>
                  </div>
                </Link>
              ))}
            </div>
          )}
          <Outlet />
        </Auth>
      </div>
    </Content>
  );
};
