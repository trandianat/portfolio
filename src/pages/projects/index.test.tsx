import { FunctionComponent, PropsWithChildren, ReactNode } from 'react';
import { useLocation } from 'react-router-dom';
import { create } from 'react-test-renderer';
import { Projects } from 'pages/projects';

const mockSignOut = jest.fn();
jest.mock('@aws-amplify/ui-react', () => ({
  withAuthenticator:
    (Auth: FunctionComponent<PropsWithChildren<{}>>) =>
    ({ children }: { children: ReactNode }) =>
      <Auth>{children}</Auth>,
  useAuthenticator: () => ({ signOut: mockSignOut }),
}));

jest.mock('react-router-dom', () => ({
  Link: ({ children, to }: { children: ReactNode; to: string }) => (
    <a href={to}>{children}</a>
  ),
  Outlet: () => <div>Outlet</div>,
  useLocation: jest.fn(),
}));

describe('Projects', () => {
  const mockUseLocation = useLocation as jest.Mock;
  mockUseLocation.mockReturnValue({ pathname: '/' });

  it('should match the snapshot', () => {
    expect(create(<Projects />)).toMatchSnapshot();
  });
});
