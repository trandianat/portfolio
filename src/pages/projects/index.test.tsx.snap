// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Projects should match the snapshot 1`] = `
[
  .emotion-0 {
  background-color: var(--white);
  padding: 48px 0;
  width: 100%;
}

.emotion-1 {
  margin: 0 auto;
  width: var(--desktop);
}

@media (max-width: 950px) {
  .emotion-1 {
    width: 650px;
  }
}

@media (max-width: 700px) {
  .emotion-1 {
    width: 400px;
  }
}

@media (max-width: 450px) {
  .emotion-1 {
    width: 80%;
  }
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 36px;
}

.emotion-3 {
  background-color: var(--black);
  height: 2px;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
}

@media (max-width: 700px) {
  .emotion-4 {
    -webkit-flex-direction: column-reverse;
    -ms-flex-direction: column-reverse;
    flex-direction: column-reverse;
    gap: 24px;
  }
}

@media (max-width: 360px) {
  .emotion-4 h1 {
    line-height: 72px;
    margin-bottom: 12px;
  }
}

.emotion-5 {
  background-color: var(--black);
  border-radius: 50%;
  color: var(--white);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 2rem;
  height: 100px;
  letter-spacing: 3px;
  min-height: 100px;
  min-width: 100px;
  width: 100px;
}

.emotion-5 .content {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 6px;
  margin: auto;
}

.emotion-5 .dot {
  background-color: var(--white);
  border-radius: 50%;
  height: 4px;
  margin: 0 auto;
  width: 4px;
}

.emotion-5 .line {
  background-color: var(--white);
  height: 1px;
  margin: 0 auto;
  width: 24px;
}

@media (max-width: 700px) {
  .emotion-5 {
    margin: 0 auto;
  }
}

@media (min-width: 951px) {
  .emotion-6 {
    display: none;
  }
}

.emotion-7 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  gap: 4px;
  vertical-align: bottom;
}

<div
    className="emotion-0"
  >
    <div
      className="emotion-1"
    >
      <div
        className="emotion-2"
      >
        <div
          className="emotion-3"
        />
        <div
          className="emotion-4"
        >
          <div>
            <h1>
              Diana Tran
            </h1>
            <p
              className="sans"
            >
              Software Engineer
            </p>
          </div>
          <div
            className="emotion-5"
          >
            <div
              className="content"
            >
              <div
                className="dot"
              />
              <div>
                DT
              </div>
              <div
                className="line"
              />
            </div>
          </div>
        </div>
        <p>
          Detail-oriented documentation expert, eager to fail fast, learn, and iterate. Experienced in design-thinking ideation and building seamless, accessible, and confidence-boosting user experiences. Well versed in the software development lifecycle including architecture and system design, testing, and monitoring.
           
          <span
            className="emotion-6"
          >
            <a
              className="emotion-7"
              href="https://www.linkedin.com/in/trandianat"
              target="_blank"
            >
              LinkedIn
              <svg
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5 3c-1.093 0-2 .907-2 2v14c0 1.093.907 2 2 2h14c1.093 0 2-.907 2-2v-7h-2v7H5V5h7V3H5zm9 0v2h3.586l-9.293 9.293 1.414 1.414L19 6.414V10h2V3h-7z"
                />
              </svg>
            </a>
             
            &
             
            <a
              className="emotion-7"
              href="https://github.com/trandianat"
              target="_blank"
            >
              Github
              <svg
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5 3c-1.093 0-2 .907-2 2v14c0 1.093.907 2 2 2h14c1.093 0 2-.907 2-2v-7h-2v7H5V5h7V3H5zm9 0v2h3.586l-9.293 9.293 1.414 1.414L19 6.414V10h2V3h-7z"
                />
              </svg>
            </a>
          </span>
        </p>
      </div>
    </div>
  </div>,
  .emotion-0 {
  background-color: var(--gray);
  padding: 48px 0;
  width: 100%;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.emotion-1 {
  margin: 0 auto;
  width: var(--desktop);
}

@media (max-width: 950px) {
  .emotion-1 {
    width: 650px;
  }
}

@media (max-width: 700px) {
  .emotion-1 {
    width: 400px;
  }
}

@media (max-width: 450px) {
  .emotion-1 {
    width: 80%;
  }
}

.emotion-2 {
  background-color: var(--black);
  border: 1px solid var(--black);
  box-shadow: var(--shadow);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: -48px 0 36px;
  position: -webkit-sticky;
  position: sticky;
  text-align: center;
  top: 0;
  z-index: 1;
}

.emotion-2 a {
  color: var(--white);
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  font-weight: normal;
  padding: 12px 0;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-2 a svg {
  fill: var(--white);
}

.emotion-2 a.active {
  background-color: var(--white);
  color: var(--black);
  font-weight: 500;
  pointer-events: none;
}

.emotion-2 a:hover {
  background-color: var(--white);
  color: var(--black);
}

.emotion-2 a:hover svg {
  fill: var(--black);
}

@media (max-width: 950px) {
  .emotion-2 .external {
    display: none;
  }
}

@media (max-width: 700px) {
  .emotion-2 a {
    font-size: .875rem;
  }
}

@media (max-width: 450px) {
  .emotion-2 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }

  .emotion-2 a {
    letter-spacing: 4px;
    padding: 10px 0;
  }

  .emotion-2 a:nth-of-type(-n + 2) {
    border-bottom: 1px solid var(--white);
  }
}

.emotion-3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  gap: 6px;
  vertical-align: bottom;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-5 button.amplify-button {
  margin-top: 18px;
}

.emotion-5 button.amplify-button--disabled {
  background-color: var(--dark-gray);
  border: 1px solid var(--dark-gray);
}

.emotion-5 button.amplify-button--disabled:hover {
  color: var(--white);
  cursor: default;
}

.emotion-5 button.amplify-field__show-password {
  display: none;
}

.emotion-5 div.amplify-alert--error {
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  border: 1px solid var(--black);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 12px;
  margin-top: 16px;
  padding: 12px;
}

.emotion-5 div.amplify-alert--error button.amplify-alert__dismiss {
  padding: 2px 0;
  margin-top: 0;
}

.emotion-5 fieldset.amplify-flex {
  border: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 16px;
  margin: 0;
  padding: 0;
}

.emotion-5 legend {
  font-size: 1.75rem;
}

.emotion-5 svg.amplify-loader {
  display: none;
}

.emotion-6 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  margin-bottom: 24px;
}

<div
    className="emotion-0"
  >
    <div
      className="emotion-1"
    >
      <nav
        className="sans emotion-2"
      >
        <a
          href="/"
        >
          Experience
        </a>
        <a
          href="/skills"
        >
          Skills
        </a>
        <a
          href="/projects"
        >
          Projects
        </a>
        <a
          className="external emotion-3"
          href="https://www.linkedin.com/in/trandianat"
          target="_blank"
        >
          <svg
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5 3c-1.093 0-2 .907-2 2v14c0 1.093.907 2 2 2h14c1.093 0 2-.907 2-2v-7h-2v7H5V5h7V3H5zm9 0v2h3.586l-9.293 9.293 1.414 1.414L19 6.414V10h2V3h-7z"
            />
          </svg>
          LinkedIn
        </a>
        <a
          className="external emotion-3"
          href="https://github.com/trandianat"
          target="_blank"
        >
          <svg
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5 3c-1.093 0-2 .907-2 2v14c0 1.093.907 2 2 2h14c1.093 0 2-.907 2-2v-7h-2v7H5V5h7V3H5zm9 0v2h3.586l-9.293 9.293 1.414 1.414L19 6.414V10h2V3h-7z"
            />
          </svg>
          Github
        </a>
      </nav>
      <div
        className="emotion-5"
      >
        <div
          className="emotion-6"
        >
          <a
            href="/projects"
          >
            â‰ª Back
          </a>
          <button
            onClick={[MockFunction]}
          >
            Sign out
          </button>
        </div>
        <div>
          Outlet
        </div>
      </div>
    </div>
  </div>,
]
`;
